### ISSUE TITLE
TensorFlow dropout code is wrong

### ISSUE THREAD

**Issue by YusukeSuzuki** [ilkarman/DeepLearningFrameworks#5]
Current code also uses dropout for testing.
Right code is like below.
I don't know that dropout should be used for train accuracies.

# High-level TF Example

```python
import numpy as np
import os
import sys
import tensorflow as tf
from common.params import *
from common.utils import *
```

```python
print("OS: ", sys.platform)
print("Python: ", sys.version)
print("Numpy: ", np.__version__)
print("Tensorflow: ", tf.__version__)
```

    OS:  linux
    Python:  3.6.0 (default, May  9 2017, 15:45:21) 
    [GCC 5.4.0 20160609]
    Numpy:  1.13.1
    Tensorflow:  1.3.0

```python
def create_symbol(training):
    conv1 = tf.layers.conv2d(X, filters=50, kernel_size=(3, 3), padding='same')
    relu1 = tf.nn.relu(conv1)
    conv2 = tf.layers.conv2d(relu1, filters=50, kernel_size=(3, 3), padding='same')
    relu2 = tf.nn.relu(conv2)
    pool1 = tf.layers.max_pooling2d(relu2, pool_size=(2, 2), strides=(2, 2), padding='valid')
    drop1 = tf.layers.dropout(pool1, 0.25, training=training)
    
    conv3 = tf.layers.conv2d(drop1, filters=100, kernel_size=(3, 3), padding='same')
    relu3 = tf.nn.relu(conv3)
    conv4 = tf.layers.conv2d(relu3, filters=100, kernel_size=(3, 3), padding='same')
    relu4 = tf.nn.relu(conv4)
    pool2 = tf.layers.max_pooling2d(relu4, pool_size=(2, 2), strides=(2, 2), padding='valid')
    drop2 = tf.layers.dropout(pool2, 0.25, training=training)
    
    flatten = tf.reshape(drop2, shape=[-1, 100*8*8])
    fc1 = tf.layers.dense(flatten, 512, activation=tf.nn.relu)
    drop4 = tf.layers.dropout(fc1, 0.5, training=training)
    logits = tf.layers.dense(drop4, N_CLASSES, name='output')
    return logits
```

```python
def init_model(m):
    # Single-class labels, don't need dense one-hot
    # Expects unscaled logits, not output of tf.nn.softmax
    xentropy = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=m, labels=y)
    loss = tf.reduce_mean(xentropy)
    optimizer = tf.train.MomentumOptimizer(learning_rate=LR, momentum=MOMENTUM)
    training_op = optimizer.minimize(loss)
    return training_op
```

```python
%%time
# Data into format for library
#x_train, x_test, y_train, y_test = mnist_for_library(channel_first=False)
x_train, x_test, y_train, y_test = cifar_for_library(channel_first=False)
print(x_train.shape, x_test.shape, y_train.shape, y_test.shape)
print(x_train.dtype, x_test.dtype, y_train.dtype, y_test.dtype)
```

    Preparing train set...
    Preparing test set...
    Done.
    (50000, 32, 32, 3) (10000, 32, 32, 3) (50000,) (10000,)
    float32 float32 int32 int32
    CPU times: user 544 ms, sys: 224 ms, total: 768 ms
    Wall time: 766 ms

```python
%%time
# Place-holders
X = tf.placeholder(tf.float32, shape=[None, 32, 32, 3])
y = tf.placeholder(tf.int32, shape=[None])
training = tf.placeholder(tf.bool)
# Initialise model
sym = create_symbol(training)
```

    CPU times: user 76 ms, sys: 4 ms, total: 80 ms
    Wall time: 78.4 ms

```python
%%time
model = init_model(sym)
init = tf.global_variables_initializer()
sess = tf.Session()
sess.run(init)
# Accuracy logging
correct = tf.nn.in_top_k(sym, y, 1)
accuracy = tf.reduce_mean(tf.cast(correct, tf.float32))
```

    CPU times: user 360 ms, sys: 632 ms, total: 992 ms
    Wall time: 1.04 s

```python
%%time
for j in range(EPOCHS):
    for data, label in yield_mb(x_train, y_train, BATCHSIZE, shuffle=True):
        sess.run(model, feed_dict={X: data, y: label, training: True})
    # Log
    acc_train = sess.run(accuracy, feed_dict={X: data, y: label, training: True})
    print(j, "Train accuracy:", acc_train)
```

    0 Train accuracy: 0.546875
    1 Train accuracy: 0.484375
    2 Train accuracy: 0.671875
    3 Train accuracy: 0.65625
    4 Train accuracy: 0.609375
    5 Train accuracy: 0.765625
    6 Train accuracy: 0.765625
    7 Train accuracy: 0.796875
    8 Train accuracy: 0.90625
    9 Train accuracy: 0.734375
    CPU times: user 1min 21s, sys: 11.9 s, total: 1min 33s
    Wall time: 1min 20s

```python
%%time
n_samples = (y_test.shape[0]//BATCHSIZE)*BATCHSIZE
y_guess = np.zeros(n_samples, dtype=np.int)
y_truth = y_test[:n_samples]
c = 0
for data, label in yield_mb(x_test, y_test, BATCHSIZE):
    pred = tf.argmax(sym,1)
    output = sess.run(pred, feed_dict={X: data, training: False})
    y_guess[c*BATCHSIZE:(c+1)*BATCHSIZE] = output
    c += 1
```

    CPU times: user 3.83 s, sys: 152 ms, total: 3.98 s
    Wall time: 3.58 s

```python
print("Accuracy: ", sum(y_guess == y_truth)/len(y_guess))
```

    Accuracy:  0.770532852564

**Comment by miguelgfierro** [comment]
We have a new winner, I got 0.78 in Teslas M60, look at [this branch](https://github.com/ilkarman/DeepLearningFrameworks/blob/bde7e1e5519af7f5a998111c9dc86b46bf2a80f7/Tensorflow_CIFAR.ipynb).

By the way what is your GPU?

**Comment by YusukeSuzuki** [comment]
My GPU is GTX 1080ti

**Comment by ilkarman** [comment]
Great! This has brought up my training accuracy from 72 to 77. Speed is a bit faster on the K80 but still slower than others (using TF 1.13)
